File: gyb_syntax_support/StmtNodes.py
from .Child import Child
from .Node import Node  # noqa: I201

STMT_NODES = [
    # labeled-stmt -> label ':' stmt
    Node('LabeledStmt', name_for_diagnostics='labeled statement', kind='Stmt',
         children=[
             Child('LabelName', kind='IdentifierToken', name_for_diagnostics='label name'),
             Child('LabelColon', kind='ColonToken'),
             Child('Statement', kind='Stmt'),
         ]),

    # expr-stmt -> expression ';'?
    Node('ExpressionStmt', name_for_diagnostics='expression', kind='Stmt',
         children=[
             Child('Expression', kind='Expr'),
         ]),

    # continue-stmt -> 'continue' label? ';'?
    Node('ContinueStmt', name_for_diagnostics="'continue' statement", kind='Stmt',
         children=[
             Child('ContinueKeyword', kind='KeywordToken', token_choices=['KeywordToken|continue']),
             Child('Label', kind='IdentifierToken', name_for_diagnostics='label',
                   is_optional=True),
         ]),

    # while-stmt -> label? ':'? 'while' condition-list code-block ';'?
    Node('WhileStmt', name_for_diagnostics="'while' statement", kind='Stmt',
         traits=['WithCodeBlock'],
         children=[
             Child('WhileKeyword', kind='KeywordToken', token_choices=['KeywordToken|while']),
             Child('Conditions', kind='ConditionElementList',
                   collection_element_name='Condition'),
             Child('Body', kind='CodeBlock'),
         ]),

    # defer-stmt -> 'defer' code-block ';'?
    Node('DeferStmt', name_for_diagnostics="'defer' statement", kind='Stmt',
         traits=['WithCodeBlock'],
         children=[
             Child('DeferKeyword', kind='KeywordToken', token_choices=['KeywordToken|defer']),
             Child('Body', kind='CodeBlock'),
         ]),

    # repeat-while-stmt -> label? ':'? 'repeat' code-block 'while' expr ';'?
    Node('RepeatWhileStmt', name_for_diagnostics="'repeat' statement", kind='Stmt',
         traits=['WithCodeBlock'],
         children=[
             Child('RepeatKeyword', kind='KeywordToken', token_choices=['KeywordToken|repeat']),
             Child('Body', kind='CodeBlock', name_for_diagnostics='body'),
             Child('WhileKeyword', kind='KeywordToken', token_choices=['KeywordToken|while']),
             Child('Condition', kind='Expr', name_for_diagnostics='condition'),
         ]),

    # guard-stmt -> 'guard' condition-list 'else' code-block ';'?
    Node('GuardStmt', name_for_diagnostics="'guard' statement", kind='Stmt',
         traits=['WithCodeBlock'],
         children=[
             Child('GuardKeyword', kind='KeywordToken', token_choices=['KeywordToken|guard']),
             Child('Conditions', kind='ConditionElementList', name_for_diagnostics='condition',
                   collection_element_name='Condition'),
             Child('ElseKeyword', kind='KeywordToken', token_choices=['KeywordToken|else']),
             Child('Body', kind='CodeBlock', name_for_diagnostics='body'),
         ]),

    Node('WhereClause', name_for_diagnostics="'where' clause", kind='Syntax',
         children=[
             Child('WhereKeyword', kind='KeywordToken', token_choices=['KeywordToken|where']),
             Child('GuardResult', kind='Expr'),
         ]),

    # for-in-stmt -> label? ':'?
    #   'for' 'try'? 'await'? 'case'? pattern 'in' expr 'where'?
    #   expr code-block ';'?
    Node('ForInStmt', name_for_diagnostics="'for' statement", kind='Stmt',
         traits=['WithCodeBlock'],
         children=[
             Child('ForKeyword', kind='KeywordToken', token_choices=['KeywordToken|for']),
             Child('TryKeyword', kind='TryToken',
                   is_optional=True),
             Child('AwaitKeyword', kind='IdentifierToken',
                   classification='Keyword',
                   token_choices=['KeywordToken|await'], is_optional=True),
             Child('CaseKeyword', kind='CaseToken',
                   is_optional=True),
             Child('Pattern', kind='Pattern'),
             Child('TypeAnnotation', kind='TypeAnnotation',
                   is_optional=True),
             Child('InKeyword', kind='KeywordToken', token_choices=['KeywordToken|in']),
             Child('SequenceExpr', kind='Expr'),
             Child('WhereClause', kind='WhereClause',
                   is_optional=True),
             Child('Body', kind='CodeBlock', name_for_diagnostics='body'),
         ]),

    # catch-clause-list -> catch-clause catch-clause-list?
    Node('CatchClauseList', name_for_diagnostics="'catch' clause",
         kind='SyntaxCollection', element='CatchClause'),

    # do-stmt -> identifier? ':'? 'do' code-block catch-clause-list ';'?
    Node('DoStmt', name_for_diagnostics="'do' statement", kind='Stmt',
         traits=['WithCodeBlock'],
         children=[
             Child('DoKeyword', kind='KeywordToken', token_choices=['KeywordToken|do']),
             Child('Body', kind='CodeBlock', name_for_diagnostics='body',),
             Child('CatchClauses', kind='CatchClauseList',
                   collection_element_name='CatchClause', is_optional=True),
         ]),

    # return-stmt -> 'return' expr? ';'?
    Node('ReturnStmt', name_for_diagnostics="'return' statement", kind='Stmt',
         children=[
             Child('ReturnKeyword', kind='KeywordToken', token_choices=['KeywordToken|return']),
             Child('Expression', kind='Expr',
                   is_optional=True),
         ]),

    # yield-stmt -> 'yield' '('? expr-list? ')'?
    Node('YieldStmt', name_for_diagnostics="'yield' statement", kind='Stmt',
         children=[
             Child('YieldKeyword', kind='KeywordToken', token_choices=['KeywordToken|yield']),
             Child('Yields', kind='Syntax',
                   node_choices=[
                       Child('YieldList', kind='YieldList'),
                       Child('SimpleYield', kind='Expr'),
                   ]),
         ]),

    Node('YieldList', name_for_diagnostics=None, kind='Syntax',
         children=[
             Child('LeftParen', kind='LeftParenToken'),
             Child('ElementList', kind='YieldExprList',
                   collection_element_name='Element'),
             Child('RightParen', kind='RightParenToken'),
         ]),

    # fallthrough-stmt -> 'fallthrough' ';'?
    Node('FallthroughStmt', name_for_diagnostics="'fallthrough' statement", kind='Stmt',
         children=[
             Child('FallthroughKeyword', kind='KeywordToken', token_choices=['KeywordToken|fallthrough']),
         ]),

    # break-stmt -> 'break' identifier? ';'?
    Node('BreakStmt', name_for_diagnostics="'break' statement", kind='Stmt',
         children=[
             Child('BreakKeyword', kind='KeywordToken', token_choices=['KeywordToken|break'], requires_trailing_space=False),
             Child('Label', kind='IdentifierToken', name_for_diagnostics='label',
                   is_optional=True),
         ]),

    # case-item-list -> case-item case-item-list?
    Node('CaseItemList', name_for_diagnostics=None, kind='SyntaxCollection',
         element='CaseItem'),

    # catch-item-list -> catch-item catch-item-list?
    Node('CatchItemList', name_for_diagnostics=None, kind='SyntaxCollection',
         element='CatchItem'),

    # condition -> expression
    #            | availability-condition
    #            | case-condition
    #            | optional-binding-condition
    Node('ConditionElement', name_for_diagnostics=None, kind='Syntax',
         traits=['WithTrailingComma'],
         children=[
             Child('Condition', kind='Syntax',
                   node_choices=[
                       Child('Expression', kind='Expr'),
                       Child('Availability', kind='AvailabilityCondition'),
                       Child('MatchingPattern',
                             kind='MatchingPatternCondition'),
                       Child('OptionalBinding',
                             kind='OptionalBindingCondition'),
                   ]),
             Child('TrailingComma', kind='CommaToken',
                   is_optional=True),
         ]),

    # availability-condition -> '#available' '(' availability-spec ')'
    Node('AvailabilityCondition', name_for_diagnostics="availability condition",
         kind='Syntax',
         children=[
             Child('AvailabilityKeyword', kind='Token', token_choices=['PoundAvailableToken', 'PoundUnavailableToken']),
             Child('LeftParen', kind='LeftParenToken'),
             Child('AvailabilitySpec', kind='AvailabilitySpecList',
                   collection_element_name='AvailabilityArgument'),
             Child('RightParen', kind='RightParenToken'),
         ]),
    Node('MatchingPatternCondition', kind='Syntax',
         name_for_diagnostics='pattern matching',
         children=[
             Child('CaseKeyword', kind='KeywordToken', token_choices=['KeywordToken|case']),
             Child('Pattern', kind='Pattern'),
             Child('TypeAnnotation', kind='TypeAnnotation',
                   is_optional=True),
             Child('Initializer', kind='InitializerClause'),
         ]),
    Node('OptionalBindingCondition', name_for_diagnostics="optional binding",
         kind='Syntax',
         children=[
             Child('BindingKeyword', kind='KeywordToken',
                   token_choices=['KeywordToken|let', 'KeywordToken|var',
                                  'KeywordToken|inout']),
             Child('Pattern', kind='Pattern'),
             Child('TypeAnnotation', kind='TypeAnnotation',
                   is_optional=True),
             Child('Initializer', kind='InitializerClause',
                   is_optional=True),
         ]),

    # condition-list -> condition
    #                 | condition ','? condition-list
    Node('ConditionElementList', name_for_diagnostics=None, kind='SyntaxCollection',
         element='ConditionElement'),

    # throw-stmt -> 'throw' expr ';'?
    Node('ThrowStmt', name_for_diagnostics="'throw' statement", kind='Stmt',
         children=[
             Child('ThrowKeyword', kind='KeywordToken', token_choices=['KeywordToken|throw']),
             Child('Expression', kind='Expr'),
         ]),

     # forget-stmt -> 'forget' expr ';'?
    Node('ForgetStmt', name_for_diagnostics="'forget' statement", kind='Stmt',
         children=[
             Child('ForgetKeyword', kind='KeywordToken', token_choices=['KeywordToken|_forget']),
             Child('Expression', kind='Expr'),
         ]),

    # catch-item -> pattern? where-clause? ','?
    Node('CatchItem', name_for_diagnostics=None, kind='Syntax',
         traits=['WithTrailingComma'],
         children=[
             Child('Pattern', kind='Pattern', is_optional=True),
             Child('WhereClause', kind='WhereClause',
                   is_optional=True),
             Child('TrailingComma', kind='CommaToken',
                   is_optional=True),
         ]),

    # catch-clause 'catch' case-item-list? code-block
    Node('CatchClause', name_for_diagnostics="'catch' clause", kind='Syntax',
         traits=['WithCodeBlock'],
         parser_function='parseCatchClause',
         children=[
             Child('CatchKeyword', kind='KeywordToken', token_choices=['KeywordToken|catch']),
             Child('CatchItems', kind='CatchItemList',
                   collection_element_name='CatchItem', is_optional=True),
             Child('Body', kind='CodeBlock'),
         ]),
]

File: CodeGeneration/Sources/SyntaxSupport/StmtNodes.swift
//===----------------------------------------------------------------------===//
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
//
//===----------------------------------------------------------------------===//

public let STMT_NODES: [Node] = [
  Node(
    kind: .availabilityCondition,
    base: .syntax,
    nameForDiagnostics: "availability condition",
    children: [
      Child(
        name: "availabilityKeyword",
        kind: .token(choices: [.token(.poundAvailable), .token(.poundUnavailable)])
      ),
      Child(
        name: "leftParen",
        kind: .token(choices: [.token(.leftParen)])
      ),
      Child(
        name: "availabilityArguments",
        deprecatedName: "availabilitySpec",
        kind: .collection(kind: .availabilityArgumentList, collectionElementName: "AvailabilityArgument")
      ),
      Child(
        name: "rightParen",
        kind: .token(choices: [.token(.rightParen)])
      ),
    ]
  ),

  Node(
    kind: .breakStmt,
    base: .stmt,
    nameForDiagnostics: "'break' statement",
    children: [
      Child(
        name: "breakKeyword",
        kind: .token(choices: [.keyword(.break)], requiresTrailingSpace: false)
      ),
      Child(
        name: "label",
        kind: .token(choices: [.token(.identifier)]),
        nameForDiagnostics: "label",
        isOptional: true
      ),
    ]
  ),

  Node(
    kind: .switchCaseItemList,
    base: .syntaxCollection,
    nameForDiagnostics: nil,
    elementChoices: [.switchCaseItem]
  ),

  Node(
    kind: .catchClauseList,
    base: .syntaxCollection,
    nameForDiagnostics: "'catch' clause",
    elementChoices: [.catchClause]
  ),

  Node(
    kind: .catchClause,
    base: .syntax,
    nameForDiagnostics: "'catch' clause",
    parserFunction: "parseCatchClause",
    traits: ["WithCodeBlock"],
    children: [
      Child(
        name: "catchKeyword",
        kind: .token(choices: [.keyword(.catch)])
      ),
      Child(
        name: "catchItems",
        kind: .collection(kind: .catchItemList, collectionElementName: "CatchItem", defaultsToEmpty: true)
      ),
      Child(
        name: "body",
        kind: .node(kind: .codeBlock)
      ),
    ]
  ),

  Node(
    kind: .catchItemList,
    base: .syntaxCollection,
    nameForDiagnostics: nil,
    elementChoices: [.catchItem]
  ),

  Node(
    kind: .catchItem,
    base: .syntax,
    nameForDiagnostics: nil,
    traits: [
      "WithTrailingComma"
    ],
    children: [
      Child(
        name: "pattern",
        kind: .node(kind: .pattern),
        isOptional: true
      ),
      Child(
        name: "whereClause",
        kind: .node(kind: .whereClause),
        isOptional: true
      ),
      Child(
        name: "trailingComma",
        kind: .token(choices: [.token(.comma)]),
        isOptional: true
      ),
    ]
  ),

  Node(
    kind: .conditionElementList,
    base: .syntaxCollection,
    nameForDiagnostics: nil,
    elementChoices: [.conditionElement]
  ),

  Node(
    kind: .conditionElement,
    base: .syntax,
    nameForDiagnostics: nil,
    traits: [
      "WithTrailingComma"
    ],
    children: [
      Child(
        name: "condition",
        kind: .nodeChoices(choices: [
          Child(
            name: "expression",
            kind: .node(kind: .expr)
          ),
          Child(
            name: "availability",
            kind: .node(kind: .availabilityCondition)
          ),
          Child(
            name: "matchingPattern",
            kind: .node(kind: .matchingPatternCondition)
          ),
          Child(
            name: "optionalBinding",
            kind: .node(kind: .optionalBindingCondition)
          ),
        ])
      ),
      Child(
        name: "trailingComma",
        kind: .token(choices: [.token(.comma)]),
        isOptional: true
      ),
    ]
  ),

  Node(
    kind: .continueStmt,
    base: .stmt,
    nameForDiagnostics: "'continue' statement",
    children: [
      Child(
        name: "continueKeyword",
        kind: .token(choices: [.keyword(.continue)])
      ),
      Child(
        name: "label",
        kind: .token(choices: [.token(.identifier)]),
        nameForDiagnostics: "label",
        isOptional: true
      ),
    ]
  ),

  Node(
    kind: .deferStmt,
    base: .stmt,
    nameForDiagnostics: "'defer' statement",
    traits: [
      "WithCodeBlock"
    ],
    children: [
      Child(
        name: "deferKeyword",
        kind: .token(choices: [.keyword(.defer)])
      ),
      Child(
        name: "body",
        kind: .node(kind: .codeBlock)
      ),
    ]
  ),

  Node(
    kind: .doStmt,
    base: .stmt,
    nameForDiagnostics: "'do' statement",
    traits: [
      "WithCodeBlock"
    ],
    children: [
      Child(
        name: "doKeyword",
        kind: .token(choices: [.keyword(.do)])
      ),
      Child(
        name: "throwsClause",
        kind: .node(kind: .throwsClause),
        documentation: "The clause specifying the type of errors thrown from the 'do' block.",
        isOptional: true
      ),
      Child(
        name: "body",
        kind: .node(kind: .codeBlock),
        nameForDiagnostics: "body"
      ),
      Child(
        name: "catchClauses",
        kind: .collection(kind: .catchClauseList, collectionElementName: "CatchClause", defaultsToEmpty: true)
      ),
    ]
  ),

  Node(
    kind: .expressionStmt,
    base: .stmt,
    nameForDiagnostics: "expression",
    children: [
      Child(
        name: "expression",
        kind: .node(kind: .expr)
      )
    ]
  ),

  Node(
    kind: .fallThroughStmt,
    base: .stmt,
    nameForDiagnostics: "'fallthrough' statement",
    children: [
      Child(
        name: "fallthroughKeyword",
        kind: .token(choices: [.keyword(.fallthrough)])
      )
    ]
  ),

  Node(
    kind: .forStmt,
    base: .stmt,
    nameForDiagnostics: "'for' statement",
    traits: [
      "WithCodeBlock"
    ],
    children: [
      Child(
        name: "forKeyword",
        kind: .token(choices: [.keyword(.for)])
      ),
      Child(
        name: "tryKeyword",
        kind: .token(choices: [.keyword(.try)]),
        isOptional: true
      ),
      Child(
        name: "awaitKeyword",
        kind: .token(choices: [.keyword(.await)]),
        isOptional: true
      ),
      Child(
        name: "caseKeyword",
        kind: .token(choices: [.keyword(.case)]),
        isOptional: true
      ),
      Child(
        name: "pattern",
        kind: .node(kind: .pattern)
      ),
      Child(
        name: "typeAnnotation",
        kind: .node(kind: .typeAnnotation),
        isOptional: true
      ),
      Child(
        name: "inKeyword",
        kind: .token(choices: [.keyword(.in)])
      ),
      Child(
        name: "sequence",
        deprecatedName: "sequenceExpr",
        kind: .node(kind: .expr)
      ),
      Child(
        name: "whereClause",
        kind: .node(kind: .whereClause),
        isOptional: true
      ),
      Child(
        name: "body",
        kind: .node(kind: .codeBlock),
        nameForDiagnostics: "body"
      ),
    ]
  ),

  Node(
    kind: .discardStmt,
    base: .stmt,
    nameForDiagnostics: "'discard' statement",
    children: [
      Child(
        name: "discardKeyword",
        kind: .token(choices: [.keyword(.discard)])
      ),
      Child(
        name: "expression",
        kind: .node(kind: .expr)
      ),
    ]
  ),

  Node(
    kind: .guardStmt,
    base: .stmt,
    nameForDiagnostics: "'guard' statement",
    traits: [
      "WithCodeBlock"
    ],
    children: [
      Child(
        name: "guardKeyword",
        kind: .token(choices: [.keyword(.guard)])
      ),
      Child(
        name: "conditions",
        kind: .collection(kind: .conditionElementList, collectionElementName: "Condition"),
        nameForDiagnostics: "condition"
      ),
      Child(
        name: "elseKeyword",
        kind: .token(choices: [.keyword(.else)])
      ),
      Child(
        name: "body",
        kind: .node(kind: .codeBlock),
        nameForDiagnostics: "body"
      ),
    ]
  ),

  Node(
    kind: .labeledStmt,
    base: .stmt,
    nameForDiagnostics: "labeled statement",
    children: [
      Child(
        name: "label",
        deprecatedName: "labelName",
        kind: .token(choices: [.token(.identifier)]),
        nameForDiagnostics: "label name"
      ),
      Child(
        name: "colon",
        deprecatedName: "labelColon",
        kind: .token(choices: [.token(.colon)])
      ),
      Child(
        name: "statement",
        kind: .node(kind: .stmt)
      ),
    ]
  ),

  Node(
    kind: .matchingPatternCondition,
    base: .syntax,
    nameForDiagnostics: "pattern matching",
    children: [
      Child(
        name: "caseKeyword",
        kind: .token(choices: [.keyword(.case)])
      ),
      Child(
        name: "pattern",
        kind: .node(kind: .pattern)
      ),
      Child(
        name: "typeAnnotation",
        kind: .node(kind: .typeAnnotation),
        isOptional: true
      ),
      Child(
        name: "initializer",
        kind: .node(kind: .initializerClause)
      ),
    ]
  ),

  Node(
    kind: .optionalBindingCondition,
    base: .syntax,
    nameForDiagnostics: "optional binding",
    children: [
      Child(
        name: "bindingSpecifier",
        deprecatedName: "bindingKeyword",
        kind: .token(choices: [
          .keyword(.let), .keyword(.var), .keyword(.inout), .keyword(._mutating), .keyword(._borrowing),
          .keyword(._consuming),
        ])
      ),
      Child(
        name: "pattern",
        kind: .node(kind: .pattern)
      ),
      Child(
        name: "typeAnnotation",
        kind: .node(kind: .typeAnnotation),
        isOptional: true
      ),
      Child(
        name: "initializer",
        kind: .node(kind: .initializerClause),
        isOptional: true
      ),
    ]
  ),

  Node(
    kind: .repeatStmt,
    base: .stmt,
    nameForDiagnostics: "'repeat' statement",
    traits: [
      "WithCodeBlock"
    ],
    children: [
      Child(
        name: "repeatKeyword",
        kind: .token(choices: [.keyword(.repeat)])
      ),
      Child(
        name: "body",
        kind: .node(kind: .codeBlock),
        nameForDiagnostics: "body"
      ),
      Child(
        name: "whileKeyword",
        kind: .token(choices: [.keyword(.while)])
      ),
      Child(
        name: "condition",
        kind: .node(kind: .expr),
        nameForDiagnostics: "condition"
      ),
    ]
  ),

  Node(
    kind: .returnStmt,
    base: .stmt,
    nameForDiagnostics: "'return' statement",
    children: [
      Child(
        name: "returnKeyword",
        kind: .token(choices: [.keyword(.return)])
      ),
      Child(
        name: "expression",
        kind: .node(kind: .expr),
        isOptional: true
      ),
    ]
  ),

  Node(
    kind: .throwStmt,
    base: .stmt,
    nameForDiagnostics: "'throw' statement",
    children: [
      Child(
        name: "throwKeyword",
        kind: .token(choices: [.keyword(.throw)])
      ),
      Child(
        name: "expression",
        kind: .node(kind: .expr)
      ),
    ]
  ),

  Node(
    kind: .whereClause,
    base: .syntax,
    nameForDiagnostics: "'where' clause",
    children: [
      Child(
        name: "whereKeyword",
        kind: .token(choices: [.keyword(.where)])
      ),
      Child(
        name: "condition",
        deprecatedName: "guardResult",
        kind: .node(kind: .expr)
      ),
    ]
  ),

  Node(
    kind: .whileStmt,
    base: .stmt,
    nameForDiagnostics: "'while' statement",
    traits: [
      "WithCodeBlock"
    ],
    children: [
      Child(
        name: "whileKeyword",
        kind: .token(choices: [.keyword(.while)])
      ),
      Child(
        name: "conditions",
        kind: .collection(kind: .conditionElementList, collectionElementName: "Condition")
      ),
      Child(
        name: "body",
        kind: .node(kind: .codeBlock)
      ),
    ]
  ),

  Node(
    kind: .yieldedExpressionsClause,
    base: .syntax,
    nameForDiagnostics: nil,
    children: [
      Child(
        name: "leftParen",
        kind: .token(choices: [.token(.leftParen)])
      ),
      Child(
        name: "elements",
        deprecatedName: "elementList",
        kind: .collection(kind: .yieldedExpressionList, collectionElementName: "Element")
      ),
      Child(
        name: "rightParen",
        kind: .token(choices: [.token(.rightParen)])
      ),
    ]
  ),

  Node(
    kind: .yieldStmt,
    base: .stmt,
    nameForDiagnostics: "'yield' statement",
    children: [
      Child(
        name: "yieldKeyword",
        kind: .token(choices: [.keyword(.yield)])
      ),
      Child(
        name: "yieldedExpressions",
        deprecatedName: "yields",
        kind: .nodeChoices(choices: [
          Child(
            name: "multiple",
            deprecatedName: "yieldList",
            kind: .node(kind: .yieldedExpressionsClause)
          ),
          Child(
            name: "single",
            deprecatedName: "simpleYield",
            kind: .node(kind: .expr)
          ),
        ])
      ),
    ]
  ),

  Node(
    kind: .thenStmt,
    base: .stmt,
    experimentalFeature: .thenStatements,
    nameForDiagnostics: "'then' statement",
    documentation: """
      A statement used to indicate the produced value from an if/switch
      expression.

       Written as:
      ```swift
      then <expr>
      ```
      """,
    children: [
      Child(
        name: "thenKeyword",
        kind: .token(choices: [.keyword(.then)])
      ),
      Child(
        name: "expression",
        kind: .node(kind: .expr)
      ),
    ]
  ),
]
